frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BaseResetBranchToRemoteAction.java

```diff
@@ -0,0 +1,188 @@
+    Option<String> branchName = getBranchName(anActionEvent);
+    Option<IGitMacheteRepository> macheteRepository = getGitMacheteRepository(anActionEvent);
+
+    if (branchName.isDefined()) {
+      if (selectedVcsRepository.isDefined()) {
+        if (macheteRepository.isDefined()) {
+          doResetToRemoteWithKeep(project, selectedVcsRepository.get(), branchName.get(), macheteRepository.get(),
+              anActionEvent);
+        } else {
+          LOG.error("Skipping the action because can't get Git Machete repository");
+          VcsNotifier.getInstance(project).notifyError(VCS_LOGGER_TITLE,
+              "Internal error occurred. For more information see IDE log file");
+        }
+      } else {
+        LOG.warn("Skipping the action because no VCS repository is selected");
+        VcsNotifier.getInstance(project).notifyWarning(VCS_LOGGER_TITLE,
+            "Skipping the action because no VCS repository is selected");
+      }
+    } else {
+      LOG.warn("Skipping the action because name of branch to reset is undefined");
+      VcsNotifier.getInstance(project).notifyError(VCS_LOGGER_TITLE,
+          "Internal error occurred. For more information see IDE log file");
+    }
+  }
+
+  protected void doResetToRemoteWithKeep(Project project, GitRepository gitRepository, String branchName,
+      IGitMacheteRepository macheteRepository, AnActionEvent anActionEvent) {
+
+    new Task.Backgroundable(project, TASK_TITLE, /* canBeCancelled */ true) {
+
+      @Override
+      public void run(ProgressIndicator indicator) {
+        LOG.debug(() -> "Resetting '${branchName}' branch");
+        try (AccessToken ignored = DvcsUtil.workingTreeChangeStarted(project, TASK_TITLE)) {
+          GitLineHandler resetHandler = new GitLineHandler(myProject, gitRepository.getRoot(), GitCommand.RESET);
+          resetHandler.addParameters("--keep");
+
+          Option<BaseGitMacheteBranch> branchOption = macheteRepository.getBranchByName(branchName);
+          assert branchOption.isDefined() : "Can't get branch '${branchName}' from Git Machete repository";
+          Option<IGitMacheteRemoteBranch> remoteTrackingBranchOption = branchOption.get().getRemoteTrackingBranch();
+          if (remoteTrackingBranchOption.isDefined()) {
+            resetHandler.addParameters(remoteTrackingBranchOption.get().getPointedCommit().getHash());
+          } else {
+            String message = "Branch '${branchName}' doesn't have remote tracking branch, so cannot be reset";
+            LOG.warn(message);
+            VcsNotifier.getInstance(project).notifyWarning(VCS_LOGGER_TITLE, message);
+          }
+
+          resetHandler.endOptions();
+
+          // Check if branch to reset is not current branch - if isn't then checkout
+          Option<String> currentBranchOption = getCurrentBranchNameIfManaged(anActionEvent);
+          if (currentBranchOption.isEmpty() || !currentBranchOption.get().equals(branchName)) {
```

Comment:
Shouldn't it be the first thing we do in this scope?

Reply:
It can be but IMHO making sure that we can reset branch (it has remote tracking branch) first is better coz we don't force user to checkout unnecessarily in this case.